### PR TITLE
Throw PreCheckException instead of setting status in PreHandleContext.

### LINKED ISSUE #5880: change preHandles to throw a PreCheckException instead of setting context status
### Problem

In existing preHandle methods, when invalid arguments are encountered, we currently set the context.status to the relevant ResponseCodeEnum and then return. Instead it would be better to throw a PreCheckException with the ResponseCodeEnum.

Also there is an implicit undocumented assumption that every implementation of TransactionHandler should have preHandle and handle methods. Methods should check payer key before doing other work.

### Solution

Update existing preHandle methods to throw the exception. Also update javadoc in TransactionHandler class to document this assumption.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by rbair23**
Closes #5880 
 - Removed from PreHandleContext the 'status' field.
 - Removed from PreHandleContext constructors that should not be public.
 - Rewrote `require` methods on PreHandleContext to keep semantics but not delegate to AccountAccess or ReadableAccountStore for validation
 - Adjusted preHandle methods accordingly now that exceptions are thrown
 - Removed some tests that were redundant / not beneficial
 - Unwrapped some fields that were Optional
 - Converted one field from account to Bytes from Optional<byte[]>
 - Removed `getKey` methods from AccountAccess and ReadableAccountStore
 - Changed semantics of getAccountByID to be Nullable etc.

### CODE REVIEW

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/handle/AdaptedMonoTransitionRunnerTest.java [hunk 1]
```diff
@@ -56,7 +56,7 @@ class AdaptedMonoTransitionRunnerTest {
     //    @Test
     //    void delegatesConsensusCreateAndTracksSuccess() {
     //        given(accessor.getFunction()).willReturn(ConsensusCreateTopic);
-    //        given(accessor.getTxn()).willReturn(mockTxn);
+    //        given(accessor.body()).willReturn(mockTxn);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/handle/AdaptedMonoTransitionRunnerTest.java:59]: Probably we forgot to fix this test while merging PBJ branch. Need not be fixed in this PR, just noting for remembering to uncomment in any of the future PRs.

File: hedera-node/hedera-consensus-service-impl/src/main/java/com/hedera/node/app/service/consensus/impl/handlers/ConsensusDeleteTopicHandler.java [hunk 2]
```diff
@@ -56,24 +57,18 @@ public ConsensusDeleteTopicHandler() {
...
-        if (adminKey.isEmpty()) {
-            context.status(ResponseCodeEnum.UNAUTHORIZED);
-            return;
-        }
-
-        context.addToReqNonPayerKeys(adminKey.get());
+        final var op = context.body().consensusDeleteTopicOrThrow();
+        // The topic ID must be present on the transaction and the topic must exist.
+        final var topic = topicStore.getTopicMetadata(op.topicID());
+        mustExist(topic, INVALID_TOPIC_ID);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-consensus-service-impl/src/main/java/com/hedera/node/app/service/consensus/impl/handlers/ConsensusDeleteTopicHandler.java:68]: Probably `validateExistsOrThrow`  instead of `mustExist` ?

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenGrantKycToAccountHandler.java [hunk 3]
```diff
@@ -48,18 +49,12 @@ public TokenGrantKycToAccountHandler() {
...
+        final var op = context.body().tokenGrantKycOrThrow();
         final var tokenMeta = tokenStore.getTokenMeta(op.tokenOrElse(TokenID.DEFAULT));
-
-        if (tokenMeta.failed()) {
-            context.status(tokenMeta.failureReason());
-            return;
-        }
-
-        tokenMeta.metadata().kycKey().ifPresent(context::addToReqNonPayerKeys);
+        tokenMeta.kycKey().ifPresent(context::requireKey);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenGrantKycToAccountHandler.java:57]: Missed 
```
if (tokenMeta == null) throw new PreCheckException(INVALID_TOKEN_ID);
```

File: hedera-node/hedera-consensus-service-impl/src/main/java/com/hedera/node/app/service/consensus/impl/handlers/ConsensusDeleteTopicHandler.java (hunk 2 above)
Comment (rbair23) [reply hedera-node/hedera-consensus-service-impl/src/main/java/com/hedera/node/app/service/consensus/impl/handlers/ConsensusDeleteTopicHandler.java:68]: 🤔

Comment (rbair23) [reply hedera-node/hedera-consensus-service-impl/src/main/java/com/hedera/node/app/service/consensus/impl/handlers/ConsensusDeleteTopicHandler.java:68]: I don't mind it other than being long-winded.

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM !

**Review by iwsimon — DISMISSED** [review]
LGTM

**Review by nathanklick — DISMISSED** [review]
Changes to the `module-info.java` and `build.gradle.kts` files are approved. No other files have been reviewed.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java [hunk 4]
```diff
@@ -17,92 +17,120 @@
...
                 accountAccess,
                 txn,
                 txn.transactionIDOrElse(TransactionID.DEFAULT).accountIDOrElse(AccountID.DEFAULT),
-                status);
+                INVALID_PAYER_ACCOUNT_ID);
     }
 
-    public PreHandleContext(
+    /** Create a new instance */
+    private PreHandleContext(
```
Comment (jsync-swirlds) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java:89]: The checks in this constructor are specific, so it's highly unlikely we would ever correctly call this private constructor with a different response code enum.  Would it not be simpler to merge this logic into the public constructor (and set the specific response code at the point the check is made) without the extra indirection?

Comment (rbair23) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java:89]: The problem is in the "createInnerContext" method, which also calls this constructor but uses a user supplied value. Initially I just hard coded the error as you suggested but when I got to the inner context I realized I needed to make this an arg

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java [hunk 5]
```diff
@@ -17,92 +17,120 @@
...
+     * @return the {@link AccountAccess}
+     */
+    @NonNull
+    public AccountAccess accountAccess() {
+        return accountAccess;
     }
 
     /**
-     * Getter for the {@link TransactionBody}
+     * Getsthe {@link TransactionBody}
```
[the remark this answers is not in the document]
Comment (rbair23) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java:118]: Dang! And it will cause the world to have to reapprove the PR. That is annoying.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java [hunk 6]
```diff
@@ -151,208 +149,238 @@ public PreHandleContext status(@NonNull final ResponseCodeEnum status) {
...
+     * @return the inner context
+     * @throws PreCheckException If the payer is not valid
      */
-    private boolean designatesContractRemoval(@NonNull final ContractID id) {
-        return id.shardNum() == 0 && id.realmNum() == 0 && id.contractNumOrElse(0L) == 0 && !id.hasEvmAddress();
+    @NonNull
+    public PreHandleContext createNestedContext(
+            @NonNull final TransactionBody nestedTxn,
+            @NonNull final AccountID payerForNested,
+            @NonNull final ResponseCodeEnum responseCode)
```
Comment (jsync-swirlds) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java:361]: This is a little odd.  It probably needs some clarifying documentation in the javadoc.  Why would I throw a different response code for an invalid payer account in this case?  I'm sure there is some code that uses it, but the javadoc should explain at least that the response code is used if the payer account is invalid or lacks a key, rather than that it's generically "used if an exception is thrown".
Basically, a caller doesn't know how to use this method correctly without reading a few levels down in code, and that's going to create confusion and incorrect usage over time.
If it's possible to elide the response code parameter, so much the better, as it's one of those things that invites incorrect usage.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java [hunk 7]
```diff
@@ -151,208 +149,238 @@ public PreHandleContext status(@NonNull final ResponseCodeEnum status) {
...
+        return innerContext;
     }
 
-    @NonNull
-    public PreHandleContext createNestedContext(
-            @NonNull final TransactionBody nestedTxn, @NonNull final AccountID payerForNested) {
-        return new PreHandleContext(accountAccess, nestedTxn, payerForNested);
+    @Override
+    public String toString() {
+        return "PreHandleContext{" + "accountAccess="
```
Comment (jsync-swirlds) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java:378]: nitpick, but this would likely read a lot better if it was a text block and `.formatted()`
e.g.
```java
return """
    PreHandleContext{accountAccess="%s", txn="%s", payer="%s", requiredNonPayerKeys="%s", \
    status="%s", innerContext="%s"}\
    """.formatted(accountAccess, txn, payer, requiredNonPayerKeys, payerKey, innerContext);
```

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleResult.java [hunk 8]
```diff
@@ -58,22 +59,23 @@ public record PreHandleResult(
 
     public PreHandleResult(
             @NonNull final PreHandleContext context,
+            @NonNull final ResponseCodeEnum status,
```
Comment (jsync-swirlds) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleResult.java:62]: Would it make sense to have this be `@Nullable` and use the context status if it is passed in as null?

File: hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Assertions.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+            throw new AssertionError("Expected " + expected + " but no exception was thrown");
+        } catch (final PreCheckException actual) {
+            if (!actual.responseCode().equals(expected)) {
+                throw new AssertionError("Expected " + expected + " but got " + actual);
+            }
+        }
+    }
+
+    /** A {@link Runnable} like interface that throws the checked {@link PreCheckException}. */
+    public interface PreCheckRunnable {
```
Comment (jsync-swirlds) [inline hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Assertions.java:45]: I think this is being used as a lambda target (if I read the QueryCheckerTest correctly).  Perhaps annotate with `@FunctionalInterface` to make that clearer (and clarify this isn't actually a `Runnable` in the threading sense).

**Review by jsync-swirlds — DISMISSED** [review]
Just a few documentation items, and nitpicks that might be good to fix in a follow-up.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java (hunk 4 above)
Comment (jsync-swirlds) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java:89]: hmmm, not sure about the inner context item; probably needs a closer look to be sure we're doing the right things there.
Not an issue for this PR, however.

File: hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Assertions.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+ */
+
+package com.hedera.node.app.spi.fixtures;
+
+import com.hedera.hapi.node.base.ResponseCodeEnum;
+import com.hedera.node.app.spi.workflows.PreCheckException;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/** A convenience class for testing with Hedera specific assertions */
+public final class Assertions {
```
Comment (kimbor) [inline hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Assertions.java:24]: I'm not convinced that this class deserves to exist. Because of this class, we can use this code in the AdminService preHandle tests:
```
        assertPreCheck(() -> subject.preHandle(context, specialFileStore), INVALID_FREEZE_TRANSACTION_BODY);
```
instead of this code:
```
        PreCheckException e = assertThrows(PreCheckException.class, () -> subject.preHandle(context, specialFileStore));
        assertSame(INVALID_FREEZE_TRANSACTION_BODY, e.responseCode());
```
so we're only saving one line of code per test and adding a bit to the cognitive overhead, since now you have to go into this class and see what it does to understand the test.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java (hunk 4 above)
Comment (rbair23) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java:89]: Agree, the inner context bothers me because a special need for Scheduled service found its way into this common API, which seems like something is wrong. But we can address it in another PR.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java (hunk 7 above)
Comment (rbair23) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java:378]: Personally I'm fine with the way it is, but in any event, this code was here before (diff just thinks it is new). I'm not sure why we even have it, TBH. I'm not a big fan of overriding toString in general, probably because I prefer debuggers over toString debuggin

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java (hunk 6 above)
Comment (rbair23) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java:361]: I agree this method is sketchy. It is only used by scheduled service, so on your scheduled service work you might end up with better insight over how this idea of nested contexts should work. Or if it can be eliminated entiirely.

File: hedera-node/hedera-mono-service/src/testFixtures/java/com/hedera/test/utils/KeyUtils.java [hunk 11]
```diff
@@ -96,6 +97,18 @@ public static List<com.hederahashgraph.api.proto.java.Key> sanityRestored(List<?
                 .toList();
     }
 
+    public static List<com.hederahashgraph.api.proto.java.Key> sanityRestored(Set<? extends HederaKey> jKeys) {
```
Comment (kimbor) [inline hedera-node/hedera-mono-service/src/testFixtures/java/com/hedera/test/utils/KeyUtils.java:100]: I've always hated this method name, ever since I first saw it. What does it mean anyway?  If it restores my sanity then it should be called restoreSanity. sanityRestored sounds like it should return a boolean (whether or not I am still insane), but then it would be clearer to call it something else `boolean isKimInsane() { return true; }`
This gripe is probably outside the scope of this PR.

File: hedera-node/hedera-mono-service/src/testFixtures/java/com/hedera/test/utils/TestFixturesKeyLookup.java [hunk 12]
```diff
@@ -53,99 +38,160 @@ public TestFixturesKeyLookup(@NonNull final ReadableStates states) {
...
-        } else {
-            final var key = account.getAccountKey();
-            final var keyResult = validateKey(key, true);
-            if (account.isReceiverSigRequired()) {
-                return keyResult;
-            } else {
-                return PRESENT_BUT_NOT_REQUIRED;
-            }
+        @Override
+        public long accountNumber() {
```
Comment (kimbor) [inline hedera-node/hedera-mono-service/src/testFixtures/java/com/hedera/test/utils/TestFixturesKeyLookup.java:74]: Method names are inconsistent, sometimes using getFoo() semantics and sometimes foo()

File: hedera-node/hedera-mono-service/src/testFixtures/java/com/hedera/test/utils/KeyUtils.java (hunk 11 above)
Comment (rbair23) [reply hedera-node/hedera-mono-service/src/testFixtures/java/com/hedera/test/utils/KeyUtils.java:100]: Agreed, I don't know what this method does either. To be addressed by another PR at some point

File: hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Assertions.java (hunk 10 above)
Comment (rbair23) [reply hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Assertions.java:24]: Maybe the name of the function should be "assertThrowsPreCheck". The example you have uses JUnit assertions, but we've already all agreed to use AssertJ, which actually makes this a little longer and more cumbersome (requires reflection to check the status!). @netopyr has another PR where he did something even more clever by extending JUnit and I really liked his approach. although it is 3-4 lines per check instead of 1. Still, his approach might be better. I've asked him in his PR to figure out the best way to combine these approaches (or just remove this one is fine too)

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java [hunk 13]
```diff
@@ -151,208 +149,238 @@ public PreHandleContext status(@NonNull final ResponseCodeEnum status) {
...
+     * @param nestedTxn the nested transaction
+     * @param payerForNested the payer for the nested transaction
+     * @param responseCode the response code to be used if a {@link PreCheckException} is thrown
+     * @return the inner context
+     * @throws PreCheckException If the payer is not valid
      */
-    private boolean designatesContractRemoval(@NonNull final ContractID id) {
-        return id.shardNum() == 0 && id.realmNum() == 0 && id.contractNumOrElse(0L) == 0 && !id.hasEvmAddress();
+    @NonNull
+    public PreHandleContext createNestedContext(
```
[the remark this answers is not in the document]
Comment (rbair23) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java:358]: I agree, this method doesn't seem to fit right. It was there before this PR so I left it alone, but I think this needs to be considered during the Scheduled service implementation.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java [hunk 14]
```diff
@@ -151,208 +149,238 @@ public PreHandleContext status(@NonNull final ResponseCodeEnum status) {
      * @return the payer key
      */
     @Nullable
-    public HederaKey getPayerKey() {
+    public HederaKey payerKey() {
```
[the remark this answers is not in the document]
Comment (rbair23) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java:152]: Correct

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleResult.java (hunk 8 above)
Comment (rbair23) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleResult.java:62]: That may be. I'll take care of this as part of the pre-handle work

File: hedera-node/hedera-schedule-service-impl/src/main/java/com/hedera/node/app/service/schedule/impl/handlers/AbstractScheduleHandler.java [hunk 15]
```diff
@@ -38,25 +39,25 @@ protected void preHandleScheduledTxn(
...
+            throws PreCheckException {
+        final var innerContext =
+                context.createNestedContext(scheduledTxn, payerForNested, UNRESOLVABLE_REQUIRED_SIGNERS);
         final HederaFunctionality scheduledFunction;
         try {
             scheduledFunction = functionOf(scheduledTxn);
         } catch (UnknownHederaFunctionality ex) {
-            innerContext.status(SCHEDULED_TRANSACTION_NOT_IN_WHITELIST);
-            return;
+            throw new PreCheckException(SCHEDULED_TRANSACTION_NOT_IN_WHITELIST);
```
[the remark this answers is not in the document]
Comment (rbair23) [reply hedera-node/hedera-schedule-service-impl/src/main/java/com/hedera/node/app/service/schedule/impl/handlers/AbstractScheduleHandler.java:50]: This is a good point, the error and response code seem wrong. If we get UnknownHederaFunctionality, something is REALLY, REALLY wrong. But since that is outside the scope of this change (i.e. even thought it doesn't look like it I am trying to avoid making extra changes :)), I encourage you to take appropriate action in your work on scheduled transaction

**Comment by jsync-swirlds** [comment]
Needs spotless run, and might need to rebase from develop to pick up the JRS test fixes merged today.
